### PR TITLE
Add Android integration for permission lifetime options.

### PR DESCRIPTION
**Pull request by goodov**
Add Android UI for permission lifetime logic.

Resolves https://github.com/brave/brave-browser/issues/14128

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by goodov** [comment]
@SergeyZhukovsky @samartnik guys, can you take a look at Android/Java/JNI part of this PR? is it good or maybe some things can be improved? I tried to keep patches size at minimum.

File: patches/components-permissions-android-BUILD.gn.patch
Comment (samartnik) [inline patches/components-permissions-android-BUILD.gn.patch:9]: nit: we should probably move these imports to `brave/android/brave_java_sources.gni`, code will look cleaner

**Review by samartnik — APPROVED** [review]
Java patches look good to me

**Review by SergeyZhukovsky — APPROVED** [review]
++

**Review by iefremov — APPROVED** [review]

File: patches/components-permissions-android-BUILD.gn.patch
Comment (goodov) [reply patches/components-permissions-android-BUILD.gn.patch:9]: tried that, but it doesn't seem right. I can't access our class from a chromium one directly, maybe the package name change and accessing it via full name will do the trick, but I don't think it will be cleaner.

File: patches/components-permissions-android-java-src-org-chromium-components-permissions-PermissionDialogDelegate.java.patch
Comment (bridiver) [inline patches/components-permissions-android-java-src-org-chromium-components-permissions-PermissionDialogDelegate.java.patch:10]: this can be done with ASM cc @samartnik

Comment (samartnik) [reply patches/components-permissions-android-java-src-org-chromium-components-permissions-PermissionDialogDelegate.java.patch:10]: here we extend JNI calls, so it won't be compilable prior to ASM manipulations

File: chromium_src/components/permissions/android/permission_dialog_delegate.h
Comment (bridiver) [inline chromium_src/components/permissions/android/permission_dialog_delegate.h:13]: is there a reason why these can't be in an anonymous namespace? Pass in what you need and return anything you need to set?

File: patches/components-permissions-android-BUILD.gn.patch
Comment (samartnik) [reply patches/components-permissions-android-BUILD.gn.patch:9]: I meant just move this part `import("//brave/components/permissions/android/java_sources.gni")`

File: patches/components-permissions-android-java-src-org-chromium-components-permissions-PermissionDialogDelegate.java.patch
Comment (samartnik) [reply patches/components-permissions-android-java-src-org-chromium-components-permissions-PermissionDialogDelegate.java.patch:10]: I just tried, JNI is not a problem, but there is still issue with casting object in `addLifetimeOptions`. We still can do it via ASM, but it's not that straightforward. I'll create a follow up issue for this.

Comment (samartnik) [reply patches/components-permissions-android-java-src-org-chromium-components-permissions-PermissionDialogDelegate.java.patch:10]: https://github.com/brave/brave-browser/issues/15243

File: patches/components-permissions-android-BUILD.gn.patch
Comment (goodov) [reply patches/components-permissions-android-BUILD.gn.patch]: got it. moved import to `brave/android/brave_java_sources.gni`

File: chromium_src/components/permissions/android/permission_dialog_delegate.h
Comment (goodov) [reply chromium_src/components/permissions/android/permission_dialog_delegate.h]: > is there a reason why these can't be in an anonymous namespace?

not really, just an easier implementation. moved to anonymous namespace.

File: android/java/proguard.flags
Comment (samartnik) [inline android/java/proguard.flags:23]: Do we need this one? We use these flags only when `BraveReflectionUtil.InvokeMethod` is used.

File: android/javatests/org/chromium/chrome/browser/BytecodeTest.java
Comment (samartnik) [inline android/javatests/org/chromium/chrome/browser/BytecodeTest.java:237]: I suppose we don't need this due to comment above.

File: android/java/proguard.flags
Comment (goodov) [reply android/java/proguard.flags:25]: not sure. if `changeMethodOwner` works without this, then it's not needed. will check it now.

File: android/java/proguard.flags
Comment (goodov) [inline android/java/proguard.flags:7]: @samartnik what about this line? added it just now. Will `changeMethodOwner` and `changeSuperName` work without proguard exclusions?

Comment (samartnik) [reply android/java/proguard.flags:7]: Yes, they will. We need exceptions here only when we invoke methods in runtime dynamically.

Comment (goodov) [reply android/java/proguard.flags:7]: oh, I think I got it.
Is all this bytecode replacement stuff done at compile/build time? And because of that the proguard flags are not needed until we want access class or invoke method via reflection at runtime.
is it correct?

Comment (samartnik) [reply android/java/proguard.flags:7]: Yep

File: android/java/proguard.flags
Comment (goodov) [reply android/java/proguard.flags]: thanks! removed proguard changes.

File: android/java/proguard.flags
Comment (goodov) [reply android/java/proguard.flags]: got it. removed unnecessary modifications to proguard.flags.

File: patches/components-permissions-android-java-src-org-chromium-components-permissions-PermissionDialogDelegate.java.patch
Comment (goodov) [reply patches/components-permissions-android-java-src-org-chromium-components-permissions-PermissionDialogDelegate.java.patch]: made all required java modifications via bytecode patches.

File: patches/components-permissions-android-permission_prompt_android.h.patch
Comment (bridiver) [inline patches/components-permissions-android-permission_prompt_android.h.patch:9]: I think you could insert these methods with the `#define PermissionCount NotUsed() { return 0 } .... size_t PermissionCount` method

File: patches/components-permissions-android-permission_dialog_delegate.cc.patch
Comment (bridiver) [inline patches/components-permissions-android-permission_dialog_delegate.cc.patch:9]: override PermissionDialogDelegate -> BravePermissionDialogDelegate in permission_prompt_android.cc an option?

File: chromium_src/components/permissions/android/permission_dialog_delegate.h
Comment (bridiver) [inline chromium_src/components/permissions/android/permission_dialog_delegate.h:13]: I thought you were converting this to anonymous?

File: patches/components-permissions-android-BUILD.gn.patch
Comment (bridiver) [inline patches/components-permissions-android-BUILD.gn.patch:9]: do we need these individual source patches for java? It seems like they are coming up a lot recently and it's not clear to me if this is strictly necessary like it is for cpp or if they can all be added in the same place somewhere else

Comment (bridiver) [reply patches/components-permissions-android-BUILD.gn.patch:9]: @AlexeyBarabash said he would look at this tomorrow

File: patches/components-permissions-android-permission_dialog_delegate.cc.patch
Comment (goodov) [reply patches/components-permissions-android-permission_dialog_delegate.cc.patch:9]: not sure how it can work, we still need to fill some data in `j_delegate_` which is created and immediately used in `PermissionDialogDelegate` constructor.
is it possible to do something in a constructor without patches? `virtual/override` will not work here.

File: chromium_src/components/permissions/android/permission_dialog_delegate.h
Comment (goodov) [reply chromium_src/components/permissions/android/permission_dialog_delegate.h:13]: yep, you're right. the second method could also be anonymized.

File: patches/components-permissions-android-permission_dialog_delegate.cc.patch
Comment (goodov) [reply patches/components-permissions-android-permission_dialog_delegate.cc.patch]: removed this patch by stubbing `Java_PermissionDialogController_createDialog` function.

File: patches/components-permissions-android-BUILD.gn.patch
Comment (AlexeyBarabash) [reply patches/components-permissions-android-BUILD.gn.patch:9]: @goodov , @bridiver,

Yes, it is possible to eliminate `brave_components_permissions_java_jni_sources` and one line of patch.
`generate_jni("jni_headers")` target in fact makes one textfile from another and does depend on java-files only, no external dependancies or other requirements. But on `jni_headers` depends the corresponding .cc file which includes the generated header.

So, here is what could be done:
1. Introduce `src/brave/components/permissions/android/BUILD.gn` with `generate_jni("jni_headers")` target
2. Remove `brave_components_permissions_java_jni_sources`  from `patches/components-permissions-android-BUILD.gn.patch` and `components/permissions/android/java_sources.gni`
3. re-use patch line at `patches/components-permissions-BUILD.gn.patch`
```
-+  import("//brave/components/permissions/sources.gni") sources += brave_components_permissions_sources deps += brave_components_permissions_deps
++  import("//brave/components/permissions/sources.gni") sources += brave_components_permissions_sources deps += brave_components_permissions_deps if (is_android) { deps += [ "//brave/components/permissions/android:jni_headers" ] }
```
4. Modify header path in `chromium_src/components/permissions/android/permission_dialog_delegate.cc`
```
-#include "components/permissions/android/jni_headers/BravePermissionDialogDelegate_jni.h"
+#include "brave/components/permissions/android/jni_headers/BravePermissionDialogDelegate_jni.h"
```

File: patches/components-permissions-android-BUILD.gn.patch
Comment (AlexeyBarabash) [inline patches/components-permissions-android-BUILD.gn.patch:17]: @bridiver
As for eliminating this line wiht `brave_components_permissions_java_sources` in patch, I recall I had errors like `class XXX not found` when my new java classes were in the different jar than Chromium's classes I tried to use. 
So maybe it is possible, but will require even more patches than that single line.

File: patches/components-permissions-android-BUILD.gn.patch
Comment (goodov) [reply patches/components-permissions-android-BUILD.gn.patch]: @AlexeyBarabash thank you!

File: components/permissions/android/BUILD.gn
Comment (AlexeyBarabash) [inline components/permissions/android/BUILD.gn:1]: nit: Brave copyright header

Comment (goodov) [reply components/permissions/android/BUILD.gn:1]: are we adding it to gn/gni? all gn files I saw didn't have one.

Comment (AlexeyBarabash) [reply components/permissions/android/BUILD.gn:1]: yes, but it is not strict checked because lint script doesn't check this.

These have the header:
https://github.com/brave/brave-core/blob/master/base/BUILD.gn
https://github.com/brave/brave-core/blob/master/android/brave_java_resources.gni

And you are right, lots doesn't. So I don't insist, but it would nice to have.
One day I believe it will be forced by lint CI check.

File: patches/components-permissions-BUILD.gn.patch
Comment (bridiver) [inline patches/components-permissions-BUILD.gn.patch:9]: just add `//brave/components/permissions/android:jni_headers` to `brave_components_permissions_deps` with an `is_android` guard. These patches are designed this way so we don't have to change them for things like this

**Review by bridiver — APPROVED** [review]
approving assuming https://github.com/brave/brave-core/pull/8466/files#r617776055 will be changed before merge

**Comment by stephendonner** [comment]
Verified `PASSED` using `nightly` build `1.25.42, Chromium 90.0.4430.85` on `Android 9; Build/PQ2A. 190305.002`

Origin-based behavior, multiple tabs (`location`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `permission.site`
4. Give permission `until I close this site`
5. tap `Allow`
6. confirm `Settings` > `Site Settings` > `Location` shows `permission.site` under `Allowed`
7. open `permission.site` in another new tab
8. close either open tab
9. confirm  `Settings` > `Site Settings` > `Location` still shows `permission.site` under `Allowed` and/or verify via tapping on the lock icon in the URL bar
10. close the remaining tab
11. confirm `Settings` > `Site Settings` > `Location` no longer lists `permission.site`

example | example | example | example | example | example
---------|---------|----------|----------|---------|----------
Screenshot_20210427-094335 | Screenshot_20210427-101513 | Screenshot_20210427-101122 | Screenshot_20210427-101251 | Screenshot_20210427-101312 | Screenshot_20210504-145139 | Screenshot_20210504-145201
</details>

Origin-based behavior, leave tab open, restart (`notifications`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `permission.site`
4. tap on `Notifications`
5. Give permission `until I close this site`
6. tap `Allow`
7. confirm `Settings` > `Site Settings` > `Notifications` shows `permission.site` under `Allowed` and/or verify via tapping on the lock icon in the URL bar
8. restart Brave
9. confirm `Settings` > `Site Settings` > `Notifications` no longer lists `permission.site`
10. on the open `permission.site tab`, tap on `Notifications` and confirm you get re-prompted for permission

example | example | example
---------|---------|----------
Screenshot (May 4, 2021 2_56_13 PM) | Screenshot (May 4, 2021 2_56_44 PM) | Screenshot (May 4, 2021 2_57_21 PM)
</details>

Origin-based behavior, close tab, open new tab (`camera + microphone`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `permission.site`
4. tap on `Camera + Microphone`
5. Give permission `until I close this site`
6. tap `Allow`
7. confirm `Settings` > `Site Settings` > `Camera` and `Microphone` show `permission.site` under `Allowed`
8. close permission.site tab
9. confirm `Settings` > `Site Settings` > `Camera` and `Microphone` no longer list `permission.site` under `Allowed` and/or verify via tapping on the lock icon in the URL bar
10. open a new tab with `permission.site`
11. on the open `permission.site` tab, tap on `Camera + Microphone` and confirm you get re-prompted for permission

example | example | example | example | example
---------|----------|----------|---------|---------
Screenshot_20210504-150206 | Screenshot_20210504-150253 | Screenshot (May 4, 2021 3_06_52 PM) | Screenshot_20210504-150236 |  Screenshot_20210504-150352
</details>


24-hours, negative test (25 hours) (`microphone`):
<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `permission.site`
4. click on `Microphone`
5. give permission `for 24 hours`
6. tap `Allow`
7. confirm `Settings` > `Site Settings` > `Microphone` show `permission.site` under `Allowed` and/or verify via tapping on the lock icon in the URL bar
8. set system clock forward by 25 hours
9. restart Brave
10. reload `permission.site`
11. tap on `Microphone` and confirm you get re-prompted for permission

example | example | example | example | example
---------|----------|---------|----------|---------

Screenshot_20210427-132607 | Screenshot_20210427-142824 | Screenshot_20210427-142817 | Screenshot_20210427-142835 | Screenshot_20210427-142901 | Screenshot_20210504-150352



</details>

1-week, positive test (6 days) (`camera`):

<summary>
<details>

1. load `brave://flags` and enable the `Permission Lifetime` flag
2. restart Brave
3. load `permission.site`
4. click on `Camera`
5. give permission for `1 week`
6. tap Allow
7. confirm `Settings` > `Site Settings` > `Camera` shows `permission.site` under `Allowed`
8. set system clock forward by 6 days
9. restart Brave
10. reload `permission.site`
11. confirm `Settings` > `Site Settings` > `Camera` still shows `permission.site` under `Allowed` and/or verify via tapping on the lock icon in the URL bar
12. tap on Camera and confirm you do NOT get re-prompted for permission

example | example | example | example | example
---------|----------|---------|----------|---------
Screenshot_20210504-142825 | Screenshot_20210504-142751 | Screenshot_20210504-142846 | Screenshot_20210504-142918 | Screenshot_20210504-142933

</details>

**Comment by stephendonner** [comment]
Just a note that I found https://github.com/brave/brave-browser/issues/15523 while testing for the uplift /cc @goodov 

We'll need to have that figured out in order to fully test this on `1.24.x` once uplifted.

**Pull request merged**
